File: apps/api/src/services/teacher_style.ts
// apps/api/src/services/teacher_style.ts
const ALLOWED_SIGNALS = new Set([
  "max_neg_dev", "max_pos_dev", "n_significant",
  "drilling_present", "drilling_improved",
  "duration_min", "mode_count", "has_piece",
]);

export type Signals = Record<string, number | boolean>;

type Tok = { kind: "number" | "ident" | "op"; text: string };

function tokenize(text: string): Tok[] {
  const tokens: Tok[] = [];
  const re = /(\d+(?:\.\d+)?)|([A-Za-z_][A-Za-z0-9_]*)|(<=|>=|==|!=|[+\-*/<>()])/g;
  for (const m of text.matchAll(re)) {
    if (m[1] !== undefined) tokens.push({ kind: "number", text: m[1] });
    else if (m[2] !== undefined) tokens.push({ kind: "ident", text: m[2] });
    else if (m[3] !== undefined) tokens.push({ kind: "op", text: m[3] });
  }
  return tokens;
}

class Cursor {
  i = 0;
  constructor(public toks: Tok[]) {}
  peek(): Tok | null { return this.toks[this.i] ?? null; }
  take(): Tok { return this.toks[this.i++]; }
}

export function evaluate(formula: string, signals: Signals): number {
  const cur = new Cursor(tokenize(formula));
  const v = expr(cur, signals);
  if (cur.peek() !== null) throw new Error(`trailing tokens in: ${formula}`);
  return v;
}

function expr(c: Cursor, s: Signals): number {
  const value = arith(c, s);
  const nxt = c.peek();
  if (nxt && nxt.kind === "ident" && nxt.text === "if") {
    c.take();
    const cond = boolExpr(c, s);
    const kw = c.peek();
    if (!kw || kw.text !== "else") throw new Error("expected 'else' in conditional");
    c.take();
    const elseValue = arith(c, s);
    return cond ? value : elseValue;
  }
  return value;
}

function boolExpr(c: Cursor, s: Signals): boolean {
  let left = boolAnd(c, s);
  while (c.peek() && c.peek()!.kind === "ident" && c.peek()!.text === "or") {
    c.take();
    const right = boolAnd(c, s);
    left = left || right;
  }
  return left;
}

function boolAnd(c: Cursor, s: Signals): boolean {
  let left = boolNot(c, s);
  while (c.peek() && c.peek()!.kind === "ident" && c.peek()!.text === "and") {
    c.take();
    const right = boolNot(c, s);
    left = left && right;
  }
  return left;
}

function boolNot(c: Cursor, s: Signals): boolean {
  const nxt = c.peek();
  if (nxt && nxt.kind === "ident" && nxt.text === "not") { c.take(); return !cmp(c, s); }
  return cmp(c, s);
}

function cmp(c: Cursor, s: Signals): boolean {
  const left = arith(c, s);
  const nxt = c.peek();
  if (nxt && ["<", "<=", ">", ">=", "==", "!="].includes(nxt.text)) {
    const op = c.take().text;
    const right = arith(c, s);
    switch (op) {
      case "<":  return left < right;
      case "<=": return left <= right;
      case ">":  return left > right;
      case ">=": return left >= right;
      case "==": return left === right;
      case "!=": return left !== right;
    }
  }
  return Boolean(left);
}

function arith(c: Cursor, s: Signals): number {
  let left = term(c, s);
  while (c.peek() && (c.peek()!.text === "+" || c.peek()!.text === "-")) {
    const op = c.take().text;
    const right = term(c, s);
    left = op === "+" ? left + right : left - right;
  }
  return left;
}

function term(c: Cursor, s: Signals): number {
  let left = factor(c, s);
  while (c.peek() && (c.peek()!.text === "*" || c.peek()!.text === "/")) {
    const op = c.take().text;
    const right = factor(c, s);
    left = op === "*" ? left * right : left / right;
  }
  return left;
}

function factor(c: Cursor, s: Signals): number {
  const nxt = c.peek();
  if (!nxt) throw new Error("unexpected end of formula");
  if (nxt.kind === "number") { c.take(); return parseFloat(nxt.text); }
  if (nxt.kind === "ident") {
    if (!ALLOWED_SIGNALS.has(nxt.text)) throw new Error(`unknown signal: ${nxt.text}`);
    c.take();
    const v = s[nxt.text];
    return typeof v === "boolean" ? (v ? 1 : 0) : Number(v);
  }
  if (nxt.text === "(") {
    c.take();
    const v = expr(c, s);
    if (!c.peek() || c.peek()!.text !== ")") throw new Error("missing close paren");
    c.take();
    return v;
  }
  throw new Error(`unexpected token: ${JSON.stringify(nxt)}`);
}

// ---- Task 13: selectClusters ----

import playbookRaw from "../lib/playbook.json";

const PRIORITY_ORDER = [
  "Technical-corrective", "Positive-encouragement", "Artifact-based",
  "Guided-discovery", "Motivational",
] as const;
const FALLBACK_PRIMARY_KEY = "Technical-corrective";
const FALLBACK_SECONDARY_KEY = "Positive-encouragement";
const CONFIDENCE_FLOOR = 0.3;

type Cluster = {
  name: string;
  language_patterns?: { register?: string; tone?: string };
  good_examples?: Array<{ text?: string; source_id?: string } | undefined>;
  when_to_use?: string[];
  triggers: { score: string };
};
type Playbook = { teaching_playbook: { clusters: Cluster[] } };

export type ClusterRef = { name: string; score: number; raw: Cluster };
export type ClusterSelection = { primary: ClusterRef; secondary: ClusterRef };

const CURLY_QUOTE_RE = /^[“”‘’„‟«»]+|[“”‘’„‟«»]+$/g;

function cleanFormula(s: string): string {
  return s.replace(CURLY_QUOTE_RE, "").trim();
}

function priorityIndex(name: string): number {
  const nameNorm = normalizeForMatch(name);
  for (let i = 0; i < PRIORITY_ORDER.length; i += 1) {
    if (nameNorm.includes(normalizeForMatch(PRIORITY_ORDER[i]))) return i;
  }
  return PRIORITY_ORDER.length;
}

function normalizeForMatch(s: string): string {
  return s.toLowerCase().replace(/-/g, "").replace(/‑/g, "");
}

function findCluster(substring: string): Cluster {
  const clusters = (playbookRaw as unknown as Playbook).teaching_playbook.clusters;
  const subNorm = normalizeForMatch(substring);
  const found = clusters.find((c) => normalizeForMatch(cleanFormula(c.name)).includes(subNorm));
  if (!found) throw new Error(`no cluster matching ${substring}`);
  return found;
}

export function selectClusters(signals: Signals): ClusterSelection {
  const clusters = (playbookRaw as unknown as Playbook).teaching_playbook.clusters;
  const scored: ClusterRef[] = clusters.map((c) => ({
    name: cleanFormula(c.name),
    score: evaluate(cleanFormula(c.triggers.score), signals),
    raw: c,
  }));
  scored.sort((a, b) => {
    if (b.score !== a.score) return b.score - a.score;
    return priorityIndex(a.name) - priorityIndex(b.name);
  });
  let primary = scored[0];
  let secondary = scored[1];
  if (primary.score < CONFIDENCE_FLOOR && secondary.score < CONFIDENCE_FLOOR) {
    const fp = findCluster(FALLBACK_PRIMARY_KEY);
    const fs = findCluster(FALLBACK_SECONDARY_KEY);
    primary = { name: cleanFormula(fp.name), score: 0, raw: fp };
    secondary = { name: cleanFormula(fs.name), score: 0, raw: fs };
  }
  return { primary, secondary };
}

File: apps/evals/shared/teacher_style.py
# apps/evals/shared/teacher_style.py
"""Cluster selection + teacher-voice prompt formatting.

Mirrors apps/api/src/services/teacher_style.ts. Both implementations evaluate
the same DSL formulas against the same signals.
"""
from __future__ import annotations
import re
from typing import Any

ALLOWED_SIGNALS = {
    "max_neg_dev", "max_pos_dev", "n_significant",
    "drilling_present", "drilling_improved",
    "duration_min", "mode_count", "has_piece",
}


class _Tokenizer:
    _TOKEN_RE = re.compile(
        r"\s*(?:"
        r"(?P<number>\d+(?:\.\d+)?)"
        r"|(?P<ident>[A-Za-z_][A-Za-z0-9_]*)"
        r"|(?P<op><=|>=|==|!=|[+\-*/<>()])"
        r")"
    )

    def __init__(self, text: str) -> None:
        self.tokens: list[tuple[str, str]] = []
        pos = 0
        while pos < len(text):
            m = self._TOKEN_RE.match(text, pos)
            if not m or m.end() == pos:
                if text[pos].isspace():
                    pos += 1
                    continue
                raise ValueError(f"unexpected character at {pos}: {text[pos]!r}")
            kind = m.lastgroup
            self.tokens.append((kind, m.group(kind)))
            pos = m.end()
        self.i = 0

    def peek(self):
        return self.tokens[self.i] if self.i < len(self.tokens) else None

    def take(self):
        tok = self.tokens[self.i]
        self.i += 1
        return tok


def evaluate(formula: str, signals: dict[str, Any]) -> float:
    tok = _Tokenizer(formula)
    value = _expr(tok, signals)
    if tok.peek() is not None:
        raise ValueError(f"unexpected trailing tokens in: {formula!r}")
    return float(value)


def _expr(tok, sig):
    value = _arith(tok, sig)
    nxt = tok.peek()
    if nxt and nxt[0] == "ident" and nxt[1] == "if":
        tok.take()
        cond = _bool_expr(tok, sig)
        kw = tok.peek()
        if not kw or kw[1] != "else":
            raise ValueError("expected 'else' in conditional")
        tok.take()
        else_value = _arith(tok, sig)
        return value if cond else else_value
    return value


def _bool_expr(tok, sig):
    left = _bool_and(tok, sig)
    while tok.peek() and tok.peek()[0] == "ident" and tok.peek()[1] == "or":
        tok.take()
        right = _bool_and(tok, sig)
        left = left or right
    return left


def _bool_and(tok, sig):
    left = _bool_not(tok, sig)
    while tok.peek() and tok.peek()[0] == "ident" and tok.peek()[1] == "and":
        tok.take()
        right = _bool_not(tok, sig)
        left = left and right
    return left


def _bool_not(tok, sig):
    nxt = tok.peek()
    if nxt and nxt[0] == "ident" and nxt[1] == "not":
        tok.take()
        return not _cmp(tok, sig)
    return _cmp(tok, sig)


def _cmp(tok, sig):
    left = _arith(tok, sig)
    nxt = tok.peek()
    if nxt and nxt[1] in ("<", "<=", ">", ">=", "==", "!="):
        op = tok.take()[1]
        right = _arith(tok, sig)
        return {
            "<": left < right, "<=": left <= right,
            ">": left > right, ">=": left >= right,
            "==": left == right, "!=": left != right,
        }[op]
    return bool(left)


def _arith(tok, sig):
    left = _term(tok, sig)
    while tok.peek() and tok.peek()[1] in ("+", "-"):
        op = tok.take()[1]
        right = _term(tok, sig)
        left = left + right if op == "+" else left - right
    return left


def _term(tok, sig):
    left = _factor(tok, sig)
    while tok.peek() and tok.peek()[1] in ("*", "/"):
        op = tok.take()[1]
        right = _factor(tok, sig)
        left = left * right if op == "*" else left / right
    return left


def _factor(tok, sig):
    nxt = tok.peek()
    if nxt is None:
        raise ValueError("unexpected end of formula")
    kind, text = nxt
    if kind == "number":
        tok.take()
        return float(text)
    if kind == "ident":
        _KEYWORDS = {"if", "else", "and", "or", "not"}
        if text in _KEYWORDS:
            raise ValueError(f"unexpected keyword in arithmetic position: {text!r}")
        if text not in ALLOWED_SIGNALS:
            raise ValueError(f"unknown signal: {text}")
        tok.take()
        v = sig[text]
        return float(v) if not isinstance(v, bool) else (1.0 if v else 0.0)
    if kind == "op" and text == "(":
        tok.take()
        v = _expr(tok, sig)
        if not tok.peek() or tok.peek()[1] != ")":
            raise ValueError("missing close paren")
        tok.take()
        return v
    raise ValueError(f"unexpected token: {nxt}")


from dataclasses import dataclass
from functools import lru_cache
from pathlib import Path
import yaml

PLAYBOOK_PATH = Path(__file__).resolve().parents[3] / "shared" / "teacher-style" / "playbook.yaml"

PRIORITY_ORDER = [
    "Technical-corrective", "Positive-encouragement", "Artifact-based",
    "Guided-discovery", "Motivational",
]
FALLBACK_PRIMARY_KEY = "Technical-corrective"
FALLBACK_SECONDARY_KEY = "Positive-encouragement"
CONFIDENCE_FLOOR = 0.3


@dataclass(frozen=True)
class ClusterRef:
    name: str
    score: float
    raw: dict


@dataclass(frozen=True)
class ClusterSelection:
    primary: ClusterRef
    secondary: ClusterRef


@lru_cache(maxsize=1)
def _load_playbook():
    return yaml.safe_load(PLAYBOOK_PATH.read_text())["teaching_playbook"]["clusters"]


def _priority_index(name: str) -> int:
    for i, key in enumerate(PRIORITY_ORDER):
        if key.lower() in name.lower():
            return i
    return len(PRIORITY_ORDER)


_QUOTE_CHARS = '"\'' + chr(0x201C) + chr(0x201D) + chr(0x2018) + chr(0x2019)


def _cluster_name(cluster: dict) -> str:
    return cluster["name"].strip(_QUOTE_CHARS)


def _normalize_name(s: str) -> str:
    return s.lower().replace("-", "").replace("‑", "")


def _find_cluster(substring: str) -> dict:
    sub_norm = _normalize_name(substring)
    for cluster in _load_playbook():
        name_norm = _normalize_name(_cluster_name(cluster))
        if sub_norm in name_norm:
            return cluster
    raise ValueError(f"no cluster matching {substring!r}")


def _formula(cluster: dict) -> str:
    raw = cluster["triggers"]["score"]
    if not isinstance(raw, str):
        return str(raw)
    # Strip ASCII and Unicode curly quotes that YAML may preserve
    return raw.strip(_QUOTE_CHARS)


def select_clusters(signals: dict[str, Any]) -> ClusterSelection:
    scored = [
        ClusterRef(name=_cluster_name(c), score=evaluate(_formula(c), signals), raw=c)
        for c in _load_playbook()
    ]
    scored.sort(key=lambda c: (-c.score, _priority_index(c.name)))
    primary, secondary = scored[0], scored[1]
    if primary.score < CONFIDENCE_FLOOR and secondary.score < CONFIDENCE_FLOOR:
        fp = _find_cluster(FALLBACK_PRIMARY_KEY)
        fs = _find_cluster(FALLBACK_SECONDARY_KEY)
        primary = ClusterRef(name=_cluster_name(fp), score=0.0, raw=fp)
        secondary = ClusterRef(name=_cluster_name(fs), score=0.0, raw=fs)
    return ClusterSelection(primary=primary, secondary=secondary)
